tests/Parse/ParseUserTest.php

```diff
@@ -81,6 +81,55 @@ public function testLoginWrongPassword()
         $user = ParseUser::logIn('asdf', 'bogus');
     }
 
+    public function testLoginWithFacebookNoId()
+    {
+        $this->setExpectedException('Parse\ParseException',
+                'Cannot log in Facebook user without an id.');
+        $user = ParseUser::logInWithFacebook(null, 'asdf');
+    }
+
+    public function testLoginWithFacebookNoAccessToken()
+    {
+        $this->setExpectedException('Parse\ParseException',
+                'Cannot log in Facebook user without an access token.');
+        $user = ParseUser::logInWithFacebook('asdf', null);
+    }
+
+    public function testLoginWithTwitterNoId()
+    {
+        $this->setExpectedException('Parse\ParseException',
+                'Cannot log in Twitter user without an id.');
+        $user = ParseUser::logInWithTwitter(null, 'asdf', 'asdf', null, 'bogus', 'bogus');
+    }
+
+    public function testLoginWithTwitterNoScreenName()
+    {
+        $this->setExpectedException('Parse\ParseException',
+                'Cannot log in Twitter user without Twitter screen name.');
+        $user = ParseUser::logInWithTwitter('asdf', null, 'asdf', null, 'bogus', 'bogus');
+    }
+
+    public function testLoginWithTwitterNoConsumerKey()
+    {
+        $this->setExpectedException('Parse\ParseException',
+                'Cannot log in Twitter user without a consumer key.');
+        $user = ParseUser::logInWithTwitter('asdf', 'asdf', null, null, 'bogus', 'bogus');
+    }
+
+    public function testLoginWithTwitterNoAuthToken()
+    {
+        $this->setExpectedException('Parse\ParseException',
+                'Cannot log in Twitter user without an auth token.');
+        $user = ParseUser::logInWithTwitter('asdf', 'asf', 'asdf', null, null, 'bogus');
+    }
+
+    public function testLoginWithTwitterNoAuthTokenSecret()
+    {
+        $this->setExpectedException('Parse\ParseException',
+                'Cannot log in Twitter user without an auth token secret.');
+        $user = ParseUser::logInWithTwitter('asdf', 'asdf', 'asdf', null, 'bogus', null);
+    }
+
```

Comment:
I'm liking the tests! The coverage is good but we should account for everything. You should add a test case for a 'bad' request to login via twitter/facebook, you can use some arbitrary string for the necessary auth values. Both tests will throw an exception, but this will help add some coverage in the process.

I noticed that ```linkWithFacebook``` and ```linkWithTwitter``` have no coverage currently, they are nearly identical in setup but should have tests written to cover them completely as well. You can simply copy and update the tests you currently have written for facebook/twitter login to handle facebook/twitter linking as well. 

Given the identical code we should probably combine the underlying preparation of 'auth' data at some point to reduce the redundancy, but that's not necessary for now; you've done quite a lot as is!